src/azul/service/responseobjects/storage_service.py

```diff
@@ -39,12 +43,192 @@ def put(self, object_key: str, data: bytes, content_type: Optional[str] = None)
 
     def get_presigned_url(self, key: str) -> str:
         return self.client.generate_presigned_url(ClientMethod='get_object',
-                                                  Params=dict(Bucket=self.__bucket_name, Key=key))
+                                                  Params=dict(Bucket=self.bucket_name, Key=key))
 
     def create_bucket(self, bucket_name: str = None):
-        self.client.create_bucket(Bucket=(bucket_name or self.__bucket_name))
-        logger.warning(f'{type(self).__name__}: Created a bucket called "{bucket_name or self.__bucket_name}"')
+        self.client.create_bucket(Bucket=(bucket_name or self.bucket_name))
+
+
+class MultipartUploadHandler:
+    """
+    S3 Multipart Upload Handler
+
+    This class is to facilitate multipart upload to S3 storage. The class itself
+    is also a context manager.
+
+    Here is the sample usage.
+
+    .. code-block:: python
+
+       handler = MultipartUploadHandler('samples.txt', 'text/plain')
+       handler.start()
+       handler.push(b'abc')
+       handler.push(b'defg')
+       # ...
+       handler.shutdown()
+
+    or
+
+    .. code-block:: python
+
+       with MultipartUploadHandler('samples.txt', 'text/plain'):
+           handler.push(b'abc')
+           handler.push(b'defg')
+           # ...
+
+    where the context pattern will automatically shutdown upon exiting the context.
+    """
+
+    def __init__(self, object_key, content_type: str):
+        self.bucket_name = config.s3_bucket
+        self.object_key = object_key
+        self.upload_id = None
+        self.handler = None
+        self.next_part_number = 1
+        self.closed = False
+        self.content_type = content_type
+        self.parts = []
+        self.futures = []
+        self.thread_pool = None
+
+    def __enter__(self):
+        return self.start()
```

Comment:
I don't like I can use this class either as a context manager or by calling the `start` and `shutdown` methods explicitly. AFAICT, we only use one approach in this project. Please inline start() and shutdown() to enforce use as a CM. 

And to preempt the "But we may want to use it in this other way in the future" response: We're not writing a general library for other developers. We're trying to achieve specific functionality at minimal cost (time, amount of code, cost of maintenance, amount of test code, amount of documentation). We don't get paid for code that offers additional flexibility or features that aren't needed for delivering the actual features we were asked to implement.